Useful error messages on connecting, unlock wallet

michaelWuensch opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

This PR has 2 main parts:
- it introduces a new feature: Unlocking wallets.
- it improves error messages

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Zap should be easy to use. Right now there are bad error messages that are not really helpful if the connection can not be established.
This PR tries to make the process of connection a lot more simpler by providing useful information.
Furthermore it is now possible to access a locked wallet and unlock it from within the app.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

It was tested with lnd 0.6.1-beta, Samsung S9, API 29

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [Contribution document](../docs/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

michaelWuensch commented:
Issues this is linked to: #17 #32 #37

maraph commented on app/src/main/java/zapsolutions/zap/util/Wallet.java:
```diff
@@ -194,6 +218,73 @@ public void run() {
+                            // To also get channels in an active state, this delay has to be quite high (5 seconds were not enough).
+                            isLNDReachable();
+                        }
+                    }, 1000);
```
So I've tested this, and since its currently set to 1000ms, I get the problem that an error is shown prior to lnd is successfully unlocked. When I put in 5 seconds it seems to be okay.
There is no possibility to register to events like wallet unlocked before the wallet is actually unlocked?

I guess what we could do to workaround this issue is to call getinfo periodically every x ms, until a certain timeout (I think its 10 seconds currently) and check for a successful result?

michaelWuensch replied:
I have now increased it to 5 seconds as a temporary solution.
I don't think unlocking is a thing that is done often with zap android. 
So I would suggest we go with this for the time being and fix it with a periodically getinfo reqest in a separate PR.
See issue #48

michaelWuensch commented on app/src/main/java/zapsolutions/zap/HomeActivity.java:
```diff
@@ -344,10 +351,52 @@ public void onWalletLoadedUpdated(boolean success, String error) {
-            Wallet.getInstance().subscribeToChannelBackup();
+
+            // ToDo: subscribeToChannelEvents() causes LND to hang, if it is called short after unlocking.
+            //Wallet.getInstance().subscribeToChannelEvents();
```
This should also be fixed when tackling issue #48

michaelWuensch commented:
A few updates to this PR:
- I have now refactored the lndconnect string parsing. The logic moved from the activity to its own class and unit test have been added for the parsing.
- I added a "retry" button that is visible when vaild connection data was provided, but still no connection can be established (LND not running, no internet connection, etc.). So now you don't have to restart the app.
- "Wrong Password" is now shown if a wrong password was used to unlock the wallet.
- A loading spinner was added for the loading screen to emphasize something is happening.
- Missing translations were added (Loading, unlock wallet)

maraph commented on app/src/main/java/zapsolutions/zap/util/Wallet.java:
```diff
@@ -155,6 +160,7 @@ public void isLNDReachable() {
 
             LightningGrpc.LightningFutureStub asyncInfoClient = LightningGrpc
```
Not particular a thing for this PR, just a question that came to my mind: I was wondering if we could keep this client as a member and use it multiple times?

michaelWuensch replied:
I tried to reuse one LightningGrpc.LightningFutureStub for all async rpc calls, but that crashed the app. The blockingStub can be reused. We might give it another try, though.

maraph commented on app/src/test/java/zapsolutions/zap/LndConnectStringParserTest.java:
```diff
@@ -0,0 +1,130 @@
+
+import static junit.framework.TestCase.assertEquals;
+
+public class LndConnectStringParserTest {
```
Nice tests 👍

maraph commented on app/src/test/java/android/util/Log.java:
```diff
@@ -0,0 +1,28 @@
+package android.util;
+
+// This is needed for the unit tests to run when the functions called include debug messages. (Mocking)
+public class Log {
```
I would suggest we add the following in our build.gradle ->
```
 testOptions { 
    unitTests.returnDefaultValues = true
  }
```
Otherwise we will have to add also a lot of other classes like this in the future whereas the gradle option is very convenient. See also: https://sites.google.com/a/android.com/tools/tech-docs/unit-testing-support#TOC-Method-...-not-mocked.-

michaelWuensch replied:
Thanks, I have changed it.

maraph commented on app/src/main/java/zapsolutions/zap/util/Wallet.java:
```diff
@@ -194,6 +218,74 @@ public void run() {
+                            // ToDo: Create a routine that retries this until successfull
+                            isLNDReachable();
+                        }
+                    }, 5000);
```
So for me only 10 seconds did the trick, if I use it with 5 seconds, I run the getinfo to early, show an error and then shortly after we are connected. :/

Maybe rather set it high first and then we can improve it with the iterative approach?

michaelWuensch replied:
Weired... on my testsystem even 1 second was enough... I increased it to 10 for now.

michaelWuensch commented on app/src/main/java/zapsolutions/zap/connection/LndConnection.java:
```diff
@@ -73,7 +73,11 @@ private void readSavedConnectionInfo() {
         byte[] certificateBytes = BaseEncoding.base64Url().decode(certificateBase64UrlString);
 
-        mSSLFactory = CustomSSLSocketFactory.create(certificateBytes);
+        try {
```
Yes, it crashed if it was unable to get a vailid certificate from the supplied bytes.
Actually I now ceck this already on the LndConnect String parser so it might not be needed again here.
The reason why I still added it here is because I don't know if it also crashes if the certificate gets invalid when it expires.

michaelWuensch commented on app/src/main/java/zapsolutions/zap/setup/ConnectRemoteNodeActivity.java:
```diff
@@ -59,11 +62,17 @@ public void onCreate(Bundle state) {
                 } catch (NullPointerException e) {
                     showError(getResources().getString(R.string.error_emptyClipboardConnect), 4000);
                 }
+                if (isClipboardContentValid) {
```
The difference is that one time it catched a NullPointer exception that occured somewhere in the verifyConnectString function, while it now only catches the Nullpointer Exception if the clipboard content is null.

michaelWuensch commented on app/src/main/res/values/strings.xml:
```diff
@@ -220,12 +224,21 @@
+    <string name="error_connection_invalid_macaroon">The macaroon is invalid.</string>
+    <string name="error_connection_invalid_host_or_port">The provided host or port is invalid.</string>
+    <string name="error_connection_invalid_macaroon2">Access denied!\n\nThe macaroon is invalid.</string>
+    <string name="error_connection_server_unreachable">The sever is not responding.\n\nPlease make sure the server is running and still has the following IP:\n\n%1$s</string>
```
done

michaelWuensch commented on app/src/main/java/zapsolutions/zap/connection/lndConnect/LndConnectStringResult.java:
```diff
@@ -0,0 +1,50 @@
+
+public class LndConnectStringResult {
+
+    private int error;
```
Nice idea! I've implemented it.

michaelWuensch commented on app/src/main/java/zapsolutions/zap/util/Wallet.java:
```diff
@@ -175,17 +181,35 @@ public void run() {
+                        broadcastWalletLoadedUpdate(false, "interrupted");
                     } catch (ExecutionException e) {
                         mConnectionCheckInProgress = false;
                         if (e.getMessage().toLowerCase().contains("unavailable")) {
```
Done

maraph commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -220,12 +224,21 @@
     <string name="error_emptyClipboardConnect">Die Zwischenablage enthält keine Verbindungsinformationen.</string>
-    <string name="error_unableToConnectToLND">Die Verbindung zum LND kann nicht hergestellt werden.</string>
     <string name="error_qr_code_result_null">QR-Code konnte nicht gelesen werden, bitte versuch es noch einmal.</string>
+    <string name="error_connection_unsupported_format">Dieses Format wird leider nicht ünterstützt.\nBitte verwende das lndconnect Format.\n\nlndconnect://...</string>
```
ünterstützt 😃

maraph commented on app/src/main/res/values-de/strings.xml:
```diff
@@ -220,12 +224,21 @@
+    <string name="error_connection_invalid_host_or_port">Der angegebene Host oder Port is ungültig.</string>
+    <string name="error_connection_invalid_macaroon2">Zugriff verweigert!\n\nDas Macaroon ist ungültig.</string>
+    <string name="error_connection_server_unreachable">Zeitüberschreitung.\nDer Server ist nicht erreichbar.\n\n- Bitte stelle sicher, dass du eine gute Internetverbindung hast\n\n- Überprüfe ob der Server läuft und unter folgender IP-Adresse erreichbar ist:\n\n%1$s</string>
+    <string name="error_connection_lnd_unavailable">Die Verbindung zum LND kann nicht hergestellt werden.\n\n- stelle sicher, dass der LND deamon auf dem server läuft.\n\n- stelle sicher, dass Port forwarding aktiviert ist. (Port: %1$s)\n\n- stelle sicher, dass das Zertifikat gültig ist.</string>
```
Maybe capitalize Server

maraph commented on app/src/main/java/zapsolutions/zap/util/Wallet.java:
```diff
@@ -1405,7 +1495,14 @@ public void unregisterWalletLoadedListener(WalletLoadedListener listener) {
     public interface WalletLoadedListener {
-        void onWalletLoadedUpdated(boolean success, String error);
+
+        int ERROR_LOCKED = 0;
```
Nice 👍
